Flaky io.camunda.zeebe.client.OAuthCredentialsProviderTest.shouldCallOauthServerOnlyOnceInMultithreadMode

**Summary**

> Try to answer the following as best as possible

- How often does the test fail?
- Does it block your work?
- Do we suspect that it is a real failure?

**Failures**

```
java.util.concurrent.CompletionException: io.grpc.StatusRuntimeException: UNKNOWN: Application error processing RPC
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
	at java.base/java.util.concurrent.CompletableFuture$BiRelay.tryFire(CompletableFuture.java:1498)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.onNext(ZeebeClientFutureImpl.java:85)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onMessage(ClientCalls.java:468)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInternal(ClientCallImpl.java:667)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:654)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.grpc.StatusRuntimeException: UNKNOWN: Application error processing RPC
	at io.grpc.Status.asRuntimeException(Status.java:533)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:481)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:574)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:72)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:742)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	... 5 more
```

**Logs**

[Test results for unit tests.zip](https://github.com/camunda/zeebe/files/14445027/Test.results.for.unit.tests.zip)

--- a/clients/java/src/test/java/io/camunda/zeebe/client/util/RecordingGatewayService.java
+++ b/clients/java/src/test/java/io/camunda/zeebe/client/util/RecordingGatewayService.java
@@ -73,19 +73,20 @@ import io.camunda.zeebe.gateway.protocol.GatewayOuterClass.UpdateJobTimeoutRespo
 import io.grpc.Status;
 import io.grpc.StatusRuntimeException;
 import io.grpc.stub.StreamObserver;
-import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.BlockingDeque;
 import java.util.concurrent.ExecutionException;
+import java.util.concurrent.LinkedBlockingDeque;
 import java.util.function.Supplier;
 
 public final class RecordingGatewayService extends GatewayImplBase {
 
-  private final List<GeneratedMessageV3> requests = new ArrayList<>();
+  private final BlockingDeque<GeneratedMessageV3> requests = new LinkedBlockingDeque<>();
 
   private final Map<Class<? extends GeneratedMessageV3>, RequestHandler> requestHandlers =
       new HashMap<>();
@@ -588,17 +589,9 @@ public final class RecordingGatewayService extends GatewayImplBase {
         });
   }
 
-  public List<GeneratedMessageV3> getRequests() {
-    return requests;
-  }
-
   @SuppressWarnings("unchecked")
-  public <T extends GeneratedMessageV3> T getRequest(final int index) {
-    return (T) requests.get(index);
-  }
-
   public <T extends GeneratedMessageV3> T getLastRequest() {
-    return getRequest(requests.size() - 1);
+    return (T) requests.getLast();
   }
 
   public <T extends GeneratedMessageV3> void addRequestHandler(

